File: app/src/main/java/de/eloc/eloc_control_panel/ng2/activities/ChangeEmailAddressActivity.java
package de.eloc.eloc_control_panel.ng2.activities;

import static de.eloc.eloc_control_panel.ng3.activities.ActivityExtensionsKt.hideKeyboard;
import static de.eloc.eloc_control_panel.ng3.activities.ActivityExtensionsKt.open;
import static de.eloc.eloc_control_panel.ng3.activities.ActivityExtensionsKt.showModalAlert;

import android.os.Bundle;
import android.text.Editable;
import android.view.MenuItem;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.appcompat.app.ActionBar;
import androidx.lifecycle.ViewModelProvider;

import de.eloc.eloc_control_panel.R;
import de.eloc.eloc_control_panel.data.UserAccountViewModel;
import de.eloc.eloc_control_panel.databinding.ActivityChangeEmailAddressBinding;
import de.eloc.eloc_control_panel.ng3.activities.ThemableActivity;
import de.eloc.eloc_control_panel.ng3.activities.VerifyEmailActivity;

public class ChangeEmailAddressActivity extends ThemableActivity {

    private ActivityChangeEmailAddressBinding binding;
    private UserAccountViewModel viewModel;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = ActivityChangeEmailAddressBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());

        viewModel = new ViewModelProvider(this).get(UserAccountViewModel.class);
        binding.currentEmailAddressTextView.setText(viewModel.getEmailAddress());

        setToolbar();
        setListeners();
        updateUI(false);
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        int id = item.getItemId();
        if (id == android.R.id.home) {
            onBackPressed();
        }
        return true;
    }

    private void updateUI(boolean locked) {
        if (locked) {
            binding.textEditLayout.setEnabled(false);
            binding.passwordLayout.setEnabled(false);
            binding.submitButton.setEnabled(false);
            binding.progressHorizontal.setVisibility(View.VISIBLE);
        } else {
            binding.textEditLayout.setEnabled(true);
            binding.passwordLayout.setEnabled(true);
            binding.submitButton.setEnabled(true);
            binding.progressHorizontal.setVisibility(View.GONE);
        }
    }

    private void setToolbar() {
        ActionBar actionBar = getSupportActionBar();
        if (actionBar != null) {
            actionBar.setDisplayHomeAsUpEnabled(true);
            actionBar.setTitle(R.string.change_email_address);
        }
    }

    private void setListeners() {
        binding.root.setOnClickListener(v -> hideKeyboard(this));
        binding.submitButton.setOnClickListener(v -> submit());
        binding.editText.addTextChangedListener(new TextInputWatcher(binding.textEditLayout));
        binding.passwordTextInput.addTextChangedListener(new TextInputWatcher(binding.passwordLayout));
    }

    private void submit() {
        hideKeyboard(this);
        binding.textEditLayout.setError(null);
        binding.passwordLayout.setError(null);

        String emailAddress = "";
        Editable editable = binding.editText.getEditableText();
        if (editable != null) {
            emailAddress = editable.toString().trim();
        }
        if (emailAddress.isEmpty()) {
            binding.textEditLayout.setError(getString(R.string.required));
            return;
        }

        String password = "";
        editable = binding.passwordTextInput.getEditableText();
        if (editable != null) {
            password = editable.toString();
        }
        if (password.isEmpty()) {
            binding.passwordLayout.setError(getString(R.string.required));
            return;
        }

        updateUI(true);
        viewModel.changeEmailAddress(emailAddress, password, this::emailAddressSubmitted);
    }

    private void emailAddressSubmitted(String error) {
        if (error == null) {
            error = "";
        }
        error = error.trim();
        String title, message;
        boolean goBack = false;
        if (error.isEmpty()) {
            title = getString(R.string.account);
            message = getString(R.string.email_address_updated);
            goBack = true;
        } else {
            title = getString(R.string.oops);
            message = error;
        }
        if (goBack) {
            showModalAlert(
                    this,
                    title,
                    message,
                    () -> open(
                            ChangeEmailAddressActivity.this,
                            VerifyEmailActivity.class,
                            true
                    )
            );
        } else {
            showModalAlert(this, title, message);
            updateUI(false);
        }
    }
}

File: app/src/main/java/de/eloc/eloc_control_panel/activities/ChangeEmailAddressActivity.kt
package de.eloc.eloc_control_panel.activities

import android.os.Bundle
import android.view.View
import androidx.lifecycle.ViewModelProvider

import de.eloc.eloc_control_panel.R
import de.eloc.eloc_control_panel.databinding.ActivityChangeEmailAddressBinding
import de.eloc.eloc_control_panel.data.UserAccountViewModel
import de.eloc.eloc_control_panel.interfaces.TextInputWatcher

class ChangeEmailAddressActivity : ThemableActivity() {

    private lateinit var binding: ActivityChangeEmailAddressBinding
    private lateinit var viewModel: UserAccountViewModel

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityChangeEmailAddressBinding.inflate(layoutInflater)
        setContentView(binding.root)

        viewModel = ViewModelProvider(this)[UserAccountViewModel::class.java]
        binding.currentEmailAddressTextView.text = viewModel.emailAddress

        setListeners()
        updateUI(false)
    }

    private fun updateUI(locked: Boolean) {
        if (locked) {
            binding.textEditLayout.isEnabled = false
            binding.passwordLayout.isEnabled = false
            binding.submitButton.isEnabled = false
            binding.progressHorizontal.visibility = View.VISIBLE
        } else {
            binding.textEditLayout.isEnabled = true
            binding.passwordLayout.isEnabled = true
            binding.submitButton.isEnabled = true
            binding.progressHorizontal.visibility = View.GONE
        }
    }

    private fun setListeners() {
        binding.root.setOnClickListener { hideKeyboard() }
        binding.submitButton.setOnClickListener { submit() }
        binding.editText.addTextChangedListener(TextInputWatcher(binding.textEditLayout))
        binding.passwordTextInput.addTextChangedListener(TextInputWatcher(binding.passwordLayout))
        binding.toolbar.setNavigationOnClickListener { goBack() }
    }

    private fun submit() {
        hideKeyboard()
        binding.textEditLayout.error = null
        binding.passwordLayout.error = null

        val emailAddress = binding.editText.editableText?.toString()?.trim() ?: ""
        if (emailAddress.isEmpty()) {
            binding.textEditLayout.error = getString(R.string.required)
            return
        }

        val password = binding.passwordTextInput.editableText?.toString() ?: ""
        if (password.isEmpty()) {
            binding.passwordLayout.error = getString(R.string.required)
            return
        }

        updateUI(true)
        viewModel.changeEmailAddress(emailAddress, password, this::emailAddressSubmitted)
    }

    private fun emailAddressSubmitted(error4: String) {
        val err = error4.trim()
        if (err.isEmpty()) {
            showModalAlert(
                getString(R.string.account),
                getString(R.string.email_address_updated)
            ) { open(VerifyEmailActivity::class.java, true) }
        } else {
            showModalAlert(getString(R.string.oops), err)
            updateUI(false)
        }
    }
}